modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/volatility/smile/function/SABRHaganVolatilityFunctionProvider.java

```diff
@@ -0,0 +1,913 @@
+/**
+ * Copyright (C) 2015 - present by OpenGamma Inc. and the OpenGamma group of companies
+ *
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.pricer.impl.volatility.smile.function;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaProperty;
+import org.joda.beans.Property;
+import org.joda.beans.impl.direct.DirectFieldsBeanBuilder;
+import org.joda.beans.impl.direct.DirectMetaBean;
+import org.joda.beans.impl.direct.DirectMetaPropertyMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.math.DoubleMath;
+import com.google.common.primitives.Doubles;
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.collect.array.DoubleArray;
+import com.opengamma.strata.math.impl.MathException;
+import com.opengamma.strata.pricer.impl.option.EuropeanVanillaOption;
+
+/**
+ * The Hagan SABR volatility function provider.
+ * <p>
+ * This class provides the functions of volatility and its sensitivity to the SABR model parameters based on the original
+ * Hagan SABR formula.  
+ * <p>
+ * Reference: Hagan, P.; Kumar, D.; Lesniewski, A. & Woodward, D. "Managing smile risk", Wilmott Magazine, 2002, September, 84-108
+ */
+@BeanDefinition
+public final class SABRHaganVolatilityFunctionProvider
+    extends VolatilityFunctionProvider<SABRFormulaData> implements ImmutableBean, Serializable {
+
+  /**
+   * Default implementation. 
+   */
+  public static final SABRHaganVolatilityFunctionProvider DEFAULT = new SABRHaganVolatilityFunctionProvider();
+
+  /* internal parameters */
+  private static final double CUTOFF_MONEYNESS = 1e-12;
+  private static final double SMALL_Z = 1e-6;
+  private static final double LARGE_NEG_Z = -1e6;
+  private static final double LARGE_POS_Z = 1e8;
+  private static final double BETA_EPS = 1e-8;
+  private static final double RHO_EPS = 1e-5;
+  private static final double RHO_EPS_NEGATIVE = 1e-8;
+  private static final double ATM_EPS = 1e-7;
+
+  @Override
+  public double getVolatility(EuropeanVanillaOption option, double forward, SABRFormulaData data) {
```

Comment:
Do we want a "EuropeanVanillaOption" object as input? In the object we don't use the Put/Call part. Moreover, in a lot of application we don't have such an object and it is created locally only for the call. This is the case for swaptions, where the strike is computed in the pricer.